CDAP-9476 create spark2 pipeline jars

albertshau opened the pull request:
Creating spark2 versions of cdap-data-pipeline and cdap-data-streams.
These will be packaged in separate directories within the artifacts
directory, and only one will be picked up as a system artifact at
runtime. The new jars just use the existing source code as source,
with different dependencies.

albertshau committed:
CDAP-9476 create spark2 pipeline jars

Creating spark2 versions of cdap-data-pipeline and cdap-data-streams.
These will be packaged in separate directories within the artifacts
directory, and only one will be picked up as a system artifact at
runtime. The new jars just use the existing source code as source,
with different dependencies.

albertshau commented:
https://builds.cask.co/browse/CDAP-DUT5779

chtyim commented on cdap-common/bin/functions.sh:
```diff
@@ -710,6 +710,9 @@ cdap_start_java() {
     # Determine SPARK_HOME
     cdap_set_spark || logecho "Could not determine SPARK_HOME! Spark support unavailable!"
+    if [[ -n ${SPARK_COMPAT} ]]; then
+      __defines+=" -Dapp.program.spark.compat=${SPARK_COMPAT}"
```
What is this for?

albertshau replied:
this is so that the right value is used in the artifact directory setting in cdap-default.xml `/opt/cdap/master/artifacts;/opt/cdap/master/artifacts/${app.program.spark.compat}`

albertshau committed:
CDAP-9476 fix units tests for data-pipeline spark2

albertshau committed:
CDAP-9476 fix spark streaming in local mode

Spark2 streaming with checkpointing expects the job jar to be in
the same place in between runs, which means we can never delete
the jar. So in local mode, create the job jar in the app temp dir,
but not in the job specific temp dir.

Also some minor fixes to get preview tests to work.

chtyim commented on cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkCompat.java:
```diff
@@ -42,9 +42,9 @@ public String getCompat() {
+    // use the value in the system property (expected in distributed)
     // otherwise, check the conf for the spark compat version (expected in standalone and unit tests)
-    String compatStr = System.getenv(Constants.SPARK_COMPAT_ENV);
+    String compatStr = System.getProperty(Constants.AppFabric.SPARK_COMPAT);
```
Can we keep both? Property > Env > cConf?

chtyim commented on cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java:
```diff
@@ -596,8 +596,15 @@ public boolean accept(String className, URL classUrl, URL classPathUrl) {
+    // in local mode, Spark Streaming with checkpointing will expect the same job jar to exist
+    // in all runs of the program. This means it can't be created in the temporary directory for the run,
+    // but must persist between runs
+    File targetDir = isLocal ? tempDir.getParentFile() : tempDir;
```
This is a bit hack to just get the parent. If really for persisting it in local mode, better pass in the `cConf` and persist it to a directory under the data directory (e.g. `data/runtime/spark/emptyJob.jar`).

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java:
```diff
@@ -55,8 +55,6 @@
   // Environment variable name for spark home
   public static final String SPARK_HOME = "SPARK_HOME";
-  // Environment variable name for spark compat version
-  public static final String SPARK_COMPAT_ENV = "SPARK_COMPAT";
```
This is used for Spark in distributed mode.

albertshau committed:
CDAP-9476 reintroduce env variable

Add back the env variable and place the spark jar in a pre-defined
place in the local data dir.

chtyim commented:
LGTM
